### PR TITLE
Add support for mid-circuit measurement program capture

### PR DESCRIPTION
**Pull request by mudit2812**
**Context:**
This PR is following up on the new program capture effort by adding support for capturing mid-circuit measurements. Currently, mid-circuit measurements can be captured, classically processed, and executed, with specific constraints, but cannot be used with `qml.cond`

**Description of the Change:**
* Add primitive for capturing `qml.measure`. This primitive's concrete implementation is basically the same as the original, where it queues a `MidMeasureMP` and returns a `MeasurementValue`, but its abstract eval returns a ~boolean~ integer tracer. Follow up work will take advantage of this.

**Benefits:**
* Mid-circuit measurements can be captured
* Arbitrary classical processing of MCMs can be captured. This includes functionality that is already available, as well as new functionality such as operations like `jnp.sin(m)`, `jnp.array([m1, m2])`, and even `jnp.max(jnp.array([m1, m2])`! All of this can be captured, but not executed.
* Using mid-circuit measurements as gate parameters, eg. `qml.RX(m, 0)`, can be captured, but not executed.
* Circuits containing mid-circuit measurements can be executed with `qml.capture.enable()`. This only works when the circuit does not classically process MCMs or applies `qml.cond` with MCMs in the predicate.
* Circuits collecting statistics on mid-circuit measurements or sequences of mid-circuit measurements can be captured and executed.

### CODE REVIEW

**Comment by mudit2812** [comment]
[sc-66783]

File: pennylane/measurements/mid_measure.py [hunk 1]
```diff
@@ -209,22 +210,55 @@ def func(x):
...
+
+    mid_measure_p = jax.core.Primitive("measure")
+
+    @mid_measure_p.def_impl
+    def _(wires, reset=False, postselect=None):
+        return _measure_impl(wires, reset=reset, postselect=postselect)
+
+    @mid_measure_p.def_abstract_eval
+    def _(*_, **__):
+        return jax.core.ShapedArray((), jax.numpy.bool_)
```
Comment (mudit2812) [inline pennylane/measurements/mid_measure.py:257]: Open question: Catalyst uses boolean tracers for abstract evals of mid-circuit measurements. However, in PennyLane, the current behaviour of measurement values is more consistent with integers. Eg, `m0 + m1` will behave like integer addition rather than a logical OR operation. I'm not sure how to reconcile this behaviour. cc @trbromley @josh146

Comment (josh146) [reply pennylane/measurements/mid_measure.py:257]: @mudit2812 I can't explain why, but `m0 + m1` works fine in Catalyst:

```python
dev = qml.device("lightning.qubit", wires=3)

@qml.qjit(autograph=True)
@qml.qnode(dev)
def circuit(x):

    def prepare_and_measure(x):
        m = jnp.zeros([2])

        for i in range(2):
            qml.RX(x, wires=i)
            m[i] = catalyst.measure(i)

        return m

    m0, m1 = prepare_and_measure(x)

    while m0 + m1 != 2:
        m0, m1 = prepare_and_measure(x)

    qml.RY(m0 + m1, wires=2)
    return {"m0": m0, "m1": m1, "probs2": qml.probs(wires=2)}
```

```pycon
>>> circuit(0.9)
{'m0': Array(1., dtype=float64),
 'm1': Array(1., dtype=float64),
 'probs2': Array([0.29192658, 0.70807342], dtype=float64)}
```

Could be because, in Python, `True + True == 2`? E.g., the casting is implicit. But may tag @dime10 in case he knows.

Comment (josh146) [reply pennylane/measurements/mid_measure.py:257]: So Catalyst _already_ behaves like PL regarding integer addition of measurements

Comment (trbromley) [reply pennylane/measurements/mid_measure.py:257]: Would be curious if there is any edge case where the behaviours differ :thinking: It looks like Python generally promotes Boolean arithmetic to integers.

Comment (mudit2812) [reply pennylane/measurements/mid_measure.py:257]: I hadn't actually executed it (since I can't), but in the jaxpr, with a boolean tracer, a logical OR and addition have the same repr for the primitive, hence my question:

```python
import pennylane as qml
import numpy as np
import jax
import jax.numpy as jnp

qml.capture.enable()

@qml.qnode(qml.device("default.qubit", wires=2, shots=10))
def f(x):
    m1 = qml.measure(0)
    m2 = qml.measure(0, reset=True)
    m3 = m1 + m2
    m4 = m1 | m2
    return qml.sample([m1, m2]), qml.expval(m4)

print(jaxpr := jax.make_jaxpr(f.func)(1.0))
```
```pycon
{ lambda ; a:f64[]. let
    b:f64[] = mul 2.0 a
    _:AbstractOperator() = RX[n_wires=1] b 0
    c:bool[] = measure[postselect=None reset=False] 0
    d:bool[] = measure[postselect=None reset=True] 0
    _:bool[] = or c d
    e:bool[] = or c d
    f:AbstractMeasurement(n_wires=2) = sample_mcm[single_mcm=False] c d
    g:AbstractMeasurement(n_wires=1) = expval_mcm[single_mcm=True] e
  in (f, g) }
```

Comment (dime10) [reply pennylane/measurements/mid_measure.py:257]: Erick figured out the reason Josh's example works is because it stores the MCM in the array `jnp.zeros(2)`, which apparently defaults to a float type 😅

Comment (dime10) [reply pennylane/measurements/mid_measure.py:257]: I found that while Python will treat the addition of Booleans like an integer addition, NumPy does not. Instead, it interprets a `+` like a logical OR, `*` like a logical AND, does not allow `-` (hinting `^` should be used instead), `^` as a logical XOR, and `/` actually produces a float :O 

Why I'm not sure, but JAX will probably follow NumPy rules on this. So I guess our decision depends on whether we want to present users with Boolean logic or integer arithmetic semantics on MCM results by default (they can always cast from one to the other). 
Personally I'm okay with either, and we can do what's most convenient for the "general" use case. @josh146 @trbromley Any thoughts here?

Comment (josh146) [reply pennylane/measurements/mid_measure.py:257]: I would suggest we use integer types by default (for now). This is partly because without _any_ processing, I would expect a measurement `m` to be of type `0` or `1` by default.

**Review by PietropaoloFrisoni — COMMENTED** [review]
Seems pretty good! Just a few questions for my own understanding

File: doc/releases/changelog-dev.md [hunk 2]
```diff
@@ -4,6 +4,9 @@
 
 <h3>New features since last release</h3>
 
+* Mid-circuit measurements can now be captured with `qml.capture` enabled.
+  [(#6015)](https://github.com/PennyLaneAI/pennylane/pull/6015)
+
```
[the remark this answers is not in the document]
Comment (mudit2812) [reply doc/releases/changelog-dev.md:9]: @trbromley @isaacdevlugt you might be able to answer this question better than me. I assume we will have a separate section in the release changelog for capture related stories. In that case, does it make more sense to split all stories into separate changelog entries or lump them into a single big entry?

File: doc/releases/changelog-dev.md [hunk 3]
```diff
@@ -145,6 +148,9 @@
 
 <h4>Community contributions 🥳</h4>
 
+* Resolved the bug in `qml.ThermalRelaxationError` where there was a typo from `tq` to `tg`.
+  [(#5988)](https://github.com/PennyLaneAI/pennylane/issues/5988)
+
```
[the remark this answers is not in the document]
Comment (mudit2812) [reply doc/releases/changelog-dev.md:153]: I'm not sure. It may be a merge conflict, I'll fix it.

[the remark this answers is not in the document]
Comment (mudit2812) [reply doc/releases/changelog-dev.md:153]: This was because the entry for this bug fix was originally under "New features since last release", which is incorrect. So I moved it to "Community Contributions" as it was an external contributor's PR.

File: doc/releases/changelog-dev.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (isaacdevlugt) [reply doc/releases/changelog-dev.md:9]: @PietropaoloFrisoni No need to worry about this at this stage 😄. For now, I would say be as verbose as you like — pump up that number of changelog entries 📈! What will most likely happen come release time is we make a bulleted list like this:

> - We captured a bunch of things.
>   (#PRnumber1)
>   (#PRnumber2)
>   ...
> 
>   The following things can be captured:
> 
>   - thing 1
>   - thing 2
>   - ...

**Review by PietropaoloFrisoni — APPROVED** [review]
LGTM

**Review by dime10 — APPROVED** [review]
Nice work 💯 

I'm guessing you abandoned `cond(m, fun)()`? 

I also vaguely remember at the start you told me that other primitives that have been implemented didn't work with measure yet because it didn't exist, and that you would need to update those. Was that just the measurement processes of mcms or do you remember if there was anything else?

**Comment by mudit2812** [comment]
> I'm guessing you abandoned `cond(m, fun)()`?

That's in a different PR.

> I also vaguely remember at the start you told me that other primitives that have been implemented didn't work with measure yet because it didn't exist, and that you would need to update those. Was that just the measurement processes of mcms or do you remember if there was anything else?

It was just the measurement processes of MCMs, which is working in this PR.
